pumpkin-inventory/src/open_container.rs

```diff
@@ -0,0 +1,69 @@
+use crate::{Container, WindowType};
+use pumpkin_world::item::ItemStack;
+use std::sync::Mutex;
+
+pub struct OpenContainer {
+    players: Vec<i32>,
+    container: Mutex<Box<dyn Container>>,
+}
+
+impl OpenContainer {
+    pub fn try_open(&self, player_id: i32) -> Option<&Mutex<Box<dyn Container>>> {
+        if !self.players.contains(&player_id) {
+            dbg!("couldn't open container");
+            return None;
+        }
+        let container = &self.container;
+        Some(container)
+    }
+
+    pub fn add_player(&mut self, player_id: i32) {
+        if !self.players.contains(&player_id) {
+            self.players.push(player_id);
+        }
+    }
+
+    pub fn remove_player(&mut self, player_id: i32) {
+        if let Some(index) = self.players.iter().enumerate().find_map(|(index, id)| {
+            if *id == player_id {
+                Some(index)
+            } else {
+                None
+            }
+        }) {
+            self.players.remove(index);
+        }
+    }
+
+    pub fn empty(player_id: i32) -> Self {
+        Self {
+            players: vec![player_id],
+            container: Mutex::new(Box::new(Chest::new())),
+        }
+    }
+
+    pub fn all_player_ids(&self) -> Vec<i32> {
+        self.players.clone()
+    }
+}
+
+struct Chest([Option<ItemStack>; 27]);
+
+impl Chest {
+    pub fn new() -> Self {
+        Self([None; 27])
+    }
+}
+impl Container for Chest {
+    fn window_type(&self) -> &'static WindowType {
+        &WindowType::Generic9x3
+    }
+
+    fn all_slots(&mut self) -> Vec<&mut Option<ItemStack>> {
```

Comment:
My thinking is that maybe a plugin will want to have their own custom rules, and hook in to item presses and such, then they'd just have to define their own struct that implements `Container` 

and the thought is that other containers that require special things be placed/taken from special places, will handle it here.

I get that `Chest` as it is now is only a wrapper, but since you can't impl a trait for a specific enum variant, it's much easier if you abstract it like this in the long run IMO